### PR TITLE
Support relative paths in local require (issue #118)

### PR DESCRIPTION
**Pull request by izaera**

### CODE REVIEW

**Comment by jbalsas** [comment]
Hey @izaera, I think this implementation is leaking module implementations, which I'm not sure wether is right or not... 

In current implementation, this module:

```javascript
define('rel-path/module-require-rel-path', ['exports', 'require', '../module7'], function (exports, require, sibling, module7) {
    'use strict';

    // Note './sibling' was not required inside the define call
    exports.resolvedDot = ( require('./sibling') );
    exports.resolvedDot2 = ( require('../module7') === module7 );
});
```

Would work with this test:

```javascript
it('should support relative paths in local require', function(done) {
    var failure = sinon.spy(function(error) {
        console.error(error);
    });
    var success = sinon.spy(function(module) {
        this.module = module;
    });

    Loader.require(['rel-path/sibling'], success, failure);
    Loader.require(['rel-path/module-require-rel-path'], success, failure);

    setTimeout(function() {
        assert.isTrue(failure.notCalled, 'Failure should not be called');
        assert.isTrue(success.calledOnce, 'Success should be called once');
        assert.isTrue(success.module.resolvedDot.default === 'sibling', 'Local require should have resolved ./ relative path');
        assert.isTrue(success.module.resolvedDot2, 'Local require should have resolved ../ relative path');

        done();
    }, 50);
});
```

Since the [Notes about synchronous require](https://github.com/requirejs/requirejs/wiki/Differences-between-the-simplified-CommonJS-wrapper-and-standard-AMD-define#notes-about-synchronous-require) we followed state that:

> On a standard AMD module you can use a synchronous require() as long as the module was already registered for that context:

Should that be avoided? What does context mean in this case?

The following implementation should take care of that and fail the previous test (as expected):

```javascript
var localRequire = function(moduleName) {
    var argc = arguments.length;

    if (argc > 1) {
        global.require.apply(global.Loader, arguments);
    } else {
        moduleName = pathResolver.resolvePath(module.name, moduleName);

        var dependencyImplementation = dependencyImplementations[module.dependencies.indexOf(moduleName)];

        if (!dependencyImplementation) {
            throw new Error('Module "' + moduleName + '" has not been loaded yet for context: ' + module.name);
        }

        return dependencyImplementation;
        }
};
```

What do you think is the right workflow, @izaera, @ipeychev?

**Comment by ipeychev** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by ipeychev** [comment]
Hey @jbalsas,

In short, I think the current implementation works as expected.

You mean that because `./sibling` is not in the list of dependencies of `'rel-path/module-require-rel-path'`, requiring it should throw an exception when required locally, don't you? 
```javascript
define('rel-path/module-require-rel-path', ['exports', 'require', '../module7'], function (exports, require, module7) {
    'use strict';

    // Note './sibling' was not required inside the define call
    exports.resolvedDot = ( require('./sibling') );
    exports.resolvedDot2 = ( require('../module7') === module7 );
});
```

However, in the test you are mentioning, the module `sibling` is required just before `module-require-rel-path`:
```javascript
Loader.require(['rel-path/sibling'], success, failure);
Loader.require(['rel-path/module-require-rel-path'], success, failure);
```

That's why it has an implementation and the local require call passes successfully.

The "correct behaviour" is not specified in the "specification", of course, but if you check what RequireJS does, it does the same as our current implementation.

Thanks,

**Comment by jbalsas** [comment]
Hey @ipeychev, the spec says:

> On a standard AMD module you can use a synchronous require() as long as the module was already registered for that context

It is true that it doesn't say exactly what should happen, but reading it I assumed you couldn't use synchronous require() if you didn't register the module in that context.

In the case I showed, the missing dependency is loaded outside the context of the module. I think the way it is opens the door for countless race conditions and weird cases where a module works in a scenario using dependencies someone else loaded, and fails in a different one when the initiator is not present. This has been always a nightmare in Liferay Portal with YUILoader...

This same code will work or fail depending on wether `sibling` is loaded before or after `module-require-rel-path` (you can simulate this throttling or delaying the first one).

```javascript
Loader.require(['rel-path/sibling'], success, failure);
Loader.require(['rel-path/module-require-rel-path'], success, failure);
```

I think I'd rather have it always fail so these issues are easily caught and code works in a predictable way.

**Comment by ipeychev** [comment]
Hey Chema,

This is correct - the code will work or not depending on if `rel-path/sibling`  is loaded before or after `module-require-rel-path`, I already tried that and it is exactly what happens with RequireJS.

I'm also pretty sure what you are assuming to be "context", is different than what is "context" is the heads of the inventors of AMD, hence this confusion.

Those are the facts - if something is not defined in the "spec", check how RequireJS works. And it does exactly that - the above code throws and exception, if not depending where `sibling` is required. Sorry to tell you that. You may check by yourself.

Thanks,

@jbalsas

**Comment by jbalsas** [comment]
Hey @ipeychev, that's fine, you're not hurting my feelings, nothing to apologize 😉 

I was just checking, the code even looks quite similar to the one in `requirejs`. I would honestly prefer a more predictable error approach here, but that's life :)

I think other that that, the rest looked good... I'll double check tomorrow before merging :)

File: src/js/script-loader.js [hunk 1]
```diff
@@ -727,18 +728,18 @@ var LoaderProtoMethods = {
...
-                                }
+                            var dependencyModule = configParser.getModules()[moduleName];
+
+                            if (!dependencyModule || !dependencyModule.implementation) {
+                                throw new Error('Module "' + moduleName + '" has not been loaded yet for context: ' + module.name);
                             }
 
-                            throw new Error('Module "' + moduleName + '" has not been loaded yet for context: ' + module.name);
-                        }
+                            return dependencyModule.implementation
```
Comment (ipeychev) [inline src/js/script-loader.js]: Here we need a semicolon :)

Comment (izaera) [inline src/js/script-loader.js]: :-)

**Comment by izaera** [comment]
Regarding the behavior of local require, I also did some checks with SystemJS and it works like our loader, i.e., if the module has been loaded previously (no matter how or from where) it works, otherwise it fails. 

I agree with @jbalsas that it does not look like a good idea and I'm also keen on failing as early as possible, but if the rest of the world does it like that I think we should follow the herd too :-(

**Comment by jbalsas** [comment]
Merged!! :)

**Pull request merged**